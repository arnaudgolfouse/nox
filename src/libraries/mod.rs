mod prelude;

use crate::vm::ffi::Library;

pub fn prelude() -> Library {
    let mut prelude = Library::new("std_prelude".to_owned());
    prelude.add_function("range", prelude::range);
    prelude.add_function("letters", prelude::letters);
    prelude.add_function("print", prelude::print);
    prelude.add_function("println", prelude::println);
    prelude.add_function("to_string", prelude::to_string);
    prelude.add_function("to_int", prelude::to_int);
    prelude
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::vm::{Value, VM};

    #[test]
    fn test_case() {
        let mut vm = VM::new();
        vm.import_all(prelude()).unwrap();
        vm.parse_top_level(
            "
            x = 0
            for i in range(1, 4)
                while i != 0
                    i -= 1
                    x += 1
                end
            end
            return x
            ",
        )
        .unwrap();
        assert_eq!(vm.run().unwrap(), Value::Int(6));

        let mut vm = VM::new();
        vm.import_all(prelude()).unwrap();
        vm.parse_top_level(
            r#"
            x = ""
            for c in letters("hello world !")
                if c == "o" or c == "e" then
                    x += c
                end
            end
            return x
            "#,
        )
        .unwrap();
        assert_eq!(vm.run().unwrap(), Value::String("eoo".to_string()));

        let mut vm = VM::new();
        vm.import_all(prelude()).unwrap();
        vm.parse_top_level(
            r#"
            a = to_string(54)   # "54"
            b = to_string(8.61) # "8.61"
            c = to_string(true) # "true"

            return a + " " + b + " " + c
            "#,
        )
        .unwrap();
        assert_eq!(vm.run().unwrap(), Value::String("54 8.61 true".to_string()));

        let mut vm = VM::new();
        vm.import_all(prelude()).unwrap();
        vm.parse_top_level(
            r#"
            a = to_int(5)    # 5
            b = to_int(8.61) # 8
            c = to_int(true) # 1
            d = to_int("-9") # -9

            return a + b + c + d
            "#,
        )
        .unwrap();
        assert_eq!(vm.run().unwrap(), Value::Int(5));
    }
}
