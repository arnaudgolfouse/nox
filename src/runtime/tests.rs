use super::*;
use crate::libraries;

#[test]
fn operations() {
    let mut vm = VM::new();
    vm.parse_top_level(
        "
x = 5                  # x = 5
y = x + 1              # y = 6
z = y * x + 8 - 2 % 2  # z = 38
return (z + 5) * -0.5  # return -21.5
        ",
    )
    .unwrap();

    assert_eq!(vm.run().unwrap(), Value::Float(-21.5));
    vm.reset();

    vm.parse_top_level(
        "
x = 2                  # x = 2
x = x ^ x              # x = 4.0
x += 5 ^ 3 % 20        # x = 9.0
return (x - 5) * -0.5  # return -2.0
        ",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Float(-2.0));
    vm.reset();

    vm.parse_top_level(
        "
return 4 + 7 + 8 - 5 - 9 # 5
        ",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(5));
    vm.reset();
}

#[test]
fn tables() {
    // manipulate fields
    let mut vm = VM::new();
    vm.parse_top_level(
        "
t = { x = 5, y = 6 }
t.y -= 9 - t.x
return t.x * t.y
        ",
    )
    .unwrap();

    assert_eq!(vm.run().unwrap(), Value::Int(10));
    vm.reset();

    // allocate / deallocate a lot + cycles
    vm.import_all(libraries::prelude()).unwrap();
    vm.parse_top_level(
        r#"
t = { x = 5, y = 6 }
t0 = t

# allocate lots
for _ in range(0, 10000)
    t0.t = { x = 1 }
    t0 = t0.t
end

t0 = t
res = 0
# deallocate lots
for _ in range(0, 10000)
    println(t0.t, "  ", t0.x, "  ", res)
    t1 = t0
    t0 = t0.t
    t1.t = nil
    res += t1.x
end

return res
        "#,
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(10004));
    vm.reset();
}

#[test]
fn control_flow() {
    let mut vm = VM::new();
    vm.parse_top_level(
        "
x = 5
y = 6
if x > y then
    if true then
        return x + y
    end
else
    if not not not false then # true
        y -= 2
    else
        y += 2
    end
end
return y",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(4));

    vm.reset();
    vm.parse_top_level(
        "
y = 1
x = 64
while x > y
    x -= 5
end
return x",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(-1));
    vm.reset();

    // testing extended instructions
    vm.parse_top_level(
        "
total = 0
y = 1
x = 64
while x > y
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
end
total += x
x = 64
while x > y
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    if x == 0 then
        break
    end
end
total += x
x = 64
while x > y
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1 x -= 1
    if x == 0 then
        continue
    end
end
total += x
return total
            ",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(0));
    vm.reset();

    vm.parse_top_level(
        "
x1 = 0
x2 = 1
i = 5
while true
    i -= 1
    x = x1 + x2
    x1 = x2
    x2 = x
    if i <= 0 then
        break
    end
end
return x2",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(8));
    vm.reset();
}

#[test]
fn for_loop() {
    let mut vm = VM::new();
    vm.parse_top_level(
        "
    fn range(a, b)
        it = a
        fn iter()
            if it == b then
                return nil
            else
                it += 1
                return it - 1
            end
        end
        return iter
    end
    x = 0
    for i in range(1, 4)
        while i != 0
            i -= 1
            x += 1
        end
    end
    return x
                ",
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(6));
    vm.reset();

    // a somewhat complex example
    vm.import_all(libraries::prelude()).unwrap();
    vm.parse_top_level(
        r#"
fn some_func(a, b)
    it = a
    fn other_func()
        if it == b then
            return nil
        else
            tot = 0
            for i in range(0, 1)      # stupid loop, only here for fun
                for i in range(1, it)
                    tot += i
                    if tot == 10 then
                        for i in range(0, 1) # same here ;)
                            tot = 0
                        end 
                        continue
                    end
                end
                it += 1
            end
            return tot
        end
    end
    return other_func
end

x = 0
for i in some_func(1, 7)
    while i != 0
        i -= 1
        x += 1
    end
end
return x
            "#,
    )
    .unwrap();
    assert_eq!(vm.run().unwrap(), Value::Int(15));
    vm.reset();
}

#[test]
fn errors() {
    let mut vm = VM::new();
    // ======== parsing errors ========
    vm.parse_top_level("0 = x").unwrap_err();
    vm.parse_top_level("fn f() break end").unwrap_err();
    vm.parse_top_level("x = 5 continue").unwrap_err();
    // ======== runtime errors ========
    // operation errors
    vm.parse_top_level("return 5 + \"hello\"").unwrap();
    vm.run().unwrap_err();
    vm.parse_top_level("return nil ^ 5.2").unwrap();
    vm.run().unwrap_err();
    vm.parse_top_level("return { x = 5 } - fn() return 1 end")
        .unwrap();
    vm.run().unwrap_err();
    vm.reset();
    // not a table
    vm.parse_top_level("x.a = 7").unwrap();
    vm.run().unwrap_err();
    vm.parse_top_level("x[789] = 9").unwrap();
    vm.run().unwrap_err();
    vm.reset();
    // not a function
    vm.parse_top_level("return 5(0, 1)").unwrap();
    vm.run().unwrap_err();
    vm.parse_top_level("x = nil return x()").unwrap();
    vm.run().unwrap_err();
    vm.reset();
}
