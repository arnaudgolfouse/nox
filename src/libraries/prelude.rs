use crate::runtime::Value;

/// Takes two `Int` arguments `a` and `b`, and creates an iterator from `a`
/// (included) to `b` (excluded).
pub fn range(args: &mut [Value]) -> Result<Value, String> {
    if args.len() != 2 {
        return Err(format!(
            "<range> : incorrect number of arguments : expected 2, got {}",
            args.len()
        ));
    }
    let from = match unsafe { args.get_unchecked(0) }.captured_value_ref() {
        Value::Int(n) => *n,
        value => return Err(format!("<range> : invalid value : {}", value)),
    };
    let to = match unsafe { args.get_unchecked(1) }.captured_value_ref() {
        Value::Int(n) => *n,
        value => return Err(format!("<range> : invalid value : {}", value)),
    };
    if to < from {
        return Ok(Value::from(|_: &mut [Value]| Ok(Value::Nil)));
    }
    let mut x = from;
    let iterator = move |_: &mut [Value]| {
        if x == to {
            Ok(Value::Nil)
        } else {
            x += 1;
            Ok(Value::Int(x - 1))
        }
    };
    Ok(Value::from(iterator))
}

/// Takes a single `String` argument, and creates an iterator over its letters.
pub fn letters(args: &mut [Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err(format!(
            "<letters> : incorrect number of arguments : expected 1, got {}",
            args.len()
        ));
    }
    let string: Vec<char> = match unsafe { args.get_unchecked(0) }.captured_value_ref() {
        Value::String(s) => s,
        value => return Err(format!("<letters> : invalid value : {}", value)),
    }
    .chars()
    .collect();
    let mut string = string.into_iter();
    let letters = move |_: &mut [Value]| {
        Ok(match string.next() {
            None => Value::Nil,
            Some(c) => Value::String(Box::from(c.encode_utf8(&mut [0; 4]) as &str)),
        })
    };
    Ok(Value::from(letters))
}

/// Print the given arguments
pub fn print(args: &mut [Value]) -> Result<Value, String> {
    for arg in args {
        print!("{}", arg)
    }
    Ok(Value::Nil)
}

/// Print the given arguments and a newline.
pub fn println(args: &mut [Value]) -> Result<Value, String> {
    for arg in args {
        print!("{}", arg)
    }
    println!();
    Ok(Value::Nil)
}

/// Takes a single argument, and tries to convert it to a `String`.
pub fn to_string(args: &mut [Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err(format!(
            "<to_string> : incorrect number of arguments : expected 1, got {}",
            args.len()
        ));
    }
    let string = match unsafe { args.get_unchecked(0) }.captured_value_ref() {
        Value::String(s) => return Ok(Value::String(s.clone())),
        Value::Int(i) => i.to_string(),
        Value::Float(f) => f.to_string(),
        Value::Bool(b) => b.to_string(),
        v => {
            return Err(format!(
                "<to_string> : impossible to convert '{}' to str",
                v
            ))
        }
    };
    Ok(Value::String(string.into_boxed_str()))
}

/// Takes a single argument, and tries to convert it to a `Int`.
pub fn to_int(args: &mut [Value]) -> Result<Value, String> {
    if args.len() != 1 {
        return Err(format!(
            "<to_int> : incorrect number of arguments : expected 1, received {}",
            args.len()
        ));
    }
    let int = match unsafe { args.get_unchecked(0) }.captured_value_ref() {
        Value::String(s) => match i64::from_str_radix(&s, 10) {
            Ok(i) => i,
            Err(err) => return Err(err.to_string()),
        },
        Value::Int(i) => *i,
        Value::Float(f) => f.floor() as i64,
        Value::Bool(b) => *b as i64,
        v => return Err(format!("<to_int> : impossible to convert '{}' to int", v)),
    };
    Ok(Value::Int(int))
}
