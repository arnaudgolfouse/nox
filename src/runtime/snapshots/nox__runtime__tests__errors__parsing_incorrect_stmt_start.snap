---
source: src/runtime/tests.rs
expression: "VM::new().parse_top_level(\"0 = x\").unwrap_err()"
---

top-level: 0
[31m  1 | [0m0 = x
[31m      ^[0m

[1;31merror[0m : unexpected token : '0'

[1;34mnote[0m : authorized in this position are 'return', 'global', 'if', 'while', 'for', 'fn', an assignement, or a function call

top-level: 0
[31m  1 | [0m0 = x
[31m          ^[0m

[1;31merror[0m : unexpected expression

[1;34mnote[0m : you can evaluate the expression by wrapping it in parenthesis -> (x)

